### ISSUE TITLE
Unexpected fail with nn.map_variable and mutable variables

### ISSUE THREAD

**Issue by Giuseppe5** [google/flax#2773]
### System information
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux
- Flax, jax, jaxlib versions (obtain with `pip show flax jax jaxlib`: Flax=0.6.3, Jax=0.3.25, Jaxlib=0.3.25
- Python version: 3.8.16
- GPU/TPU model and memory: None
- CUDA version (if applicable): None

### Problem you have encountered:
`nn.map_variable` does not work as intended when used in a module that contains mutable variables (i.e., BatchNorm).

The init phase works fine, however during a normal execution (i.e., when calling apply), the execution fails (see error below).

### What you expected to happen:
For nn.map_variable to work when calling `apply`

### Logs, error messages, etc:

> ScopeCollectionNotFound: Tried to access "kernel" from collection "params" in "/map_variables(Map_variablesConv_0)" but the collection is empty. 

### Steps to reproduce:
Whenever possible, please provide a *minimal example*. Please consider submitting it as a Colab link.

```python
class BasicConv2d(nn.Module):
  out_features: int

  @nn.compact
  def __call__(self, x, train=True):
    def sign(x):
      return jax.tree_util.tree_map(jnp.sign, x)
    Conv = nn.map_variables(nn.Conv,'params', sign, init=True)
    x = Conv(self.out_features,kernel_size=(3,3), padding='VALID')(x)
    x = nn.BatchNorm(use_running_average=not train, axis=-1)(x)
    x = nn.activation.relu(x)
    return x
```

After digging a bit, it seems the error is related to this:
https://github.com/google/flax/blob/2946b3c84cd8e069ff71408389758fa9f17684bc/flax/core/lift.py#L274-L275

Is this check here to differentiate when we're in init phase vs execution phase?
I was reading in #2486 that there is a specific method for that, but I am not sure if it could be relevant for this.

Thanks for your help!

**Comment by cgarciae** [comment]
Hey @Giuseppe5, I could not reproduce your error with the following code:

```python
import flax.linen as nn
import jax
import jax.numpy as jnp

class BasicConv2d(nn.Module):
  out_features: int

  @nn.compact
  def __call__(self, x, train=True):
    def sign(x):
      return jax.tree_util.tree_map(jnp.sign, x)
    Conv = nn.map_variables(nn.Conv,'params', sign, init=True)
    x = Conv(self.out_features,kernel_size=(3,3), padding='VALID')(x)
    x = nn.BatchNorm(use_running_average=not train, axis=-1)(x)
    x = nn.activation.relu(x)
    return x

module = BasicConv2d(32)
t, variables = module.init_with_output(jax.random.PRNGKey(0), jnp.ones((1, 32, 32, 3)), train=False)

print(jax.tree_map(jnp.shape, variables))
```
`train=True` also works. Is there anything we are doing differently?

**Comment by Giuseppe5** [comment]
```
module = BasicConv2d(32)
key, subkey = jax.random.split(jax.random.PRNGKey(0))
inp = jnp.ones((1, 32, 32, 3))
t, variables = module.init_with_output(jax.random.PRNGKey(0), inp, train=False)

state, params = variables.pop('params')

logits, updates = module.apply(
  {'params': params, 'batch_stats': state},
  x=inp, train=True, mutable=['batch_stats'])
```
Not including the model definion, this is what gives me an error, when running `module.apply`

**Comment by cgarciae** [comment]
Using `state` here is not correct, try:

```python
module = BasicConv2d(32)
key, subkey = jax.random.split(jax.random.PRNGKey(0))
inp = jnp.ones((1, 32, 32, 3))
t, variables = module.init_with_output(jax.random.PRNGKey(0), inp, train=False)

params = variables['params']
batch_stats = variables['batch_stats']

logits, updates = module.apply(
  {'params': params, 'batch_stats': batch_stats},
  x=inp, train=True, mutable=['batch_stats'])
```

**Comment by cgarciae** [comment]
I can confirm that even with the fix I proposed the error still occurs. Will look more into it.

**Comment by cgarciae** [comment]
Ok, so I figured out a solution for you problem which is using `self.is_initializing()` to define the `init` argument:

```python
import flax.linen as nn
import jax
import jax.numpy as jnp

class BasicConv2d(nn.Module):
  out_features: int

  @nn.compact
  def __call__(self, x, train=True):
    def sign(x):
      return jax.tree_util.tree_map(jnp.sign, x)
    Conv = nn.map_variables(nn.Conv, True, sign, init=self.is_initializing()) # <== HERE
    x = Conv(self.out_features, kernel_size=(3,3), padding='VALID')(x)
    x = nn.BatchNorm(use_running_average=not train, axis=-1)(x)
    x = nn.activation.relu(x)
    return x

module = BasicConv2d(32)
x = jnp.ones((1, 32, 32, 3))
t, variables = module.init_with_output(jax.random.PRNGKey(0), x, train=False)

logits, updates = module.apply(
  variables, x, train=True, mutable=['batch_stats'])

variables = variables.copy(updates)
print(jax.tree_map(jnp.shape, variables))
```
---
Some idea for the future: to solve this issue more generally, I am wondering if in `map_variables` we can change the signature of `init` to 

```python
init: Optional[bool] = None
```
and somehow make it default to `self.is_initializing()` later if not provided.

**Comment by cgarciae** [comment]
I am closing this issue. Feel free to open it if something comes up.
